homeassistant/components/mqtt/tag.py

```diff
@@ -0,0 +1,224 @@
+        config_entry,
+    )
+
+    await tag_scanner.setup()
+
+    if device_id:
+        hass.data[TAGS][device_id][discovery_id] = tag_scanner
+
+
+def async_has_tags(hass, device_id):
+    """Device has tag scanners."""
+    if TAGS not in hass.data or device_id not in hass.data[TAGS]:
+        return False
+    return hass.data[TAGS][device_id] != {}
+
+
+class MQTTTagScanner:
+    """MQTT Tag scanner."""
+
+    def __init__(self, hass, config, device_id, discovery_data, config_entry):
+        """Initialize."""
+        self._config = config
+        self._config_entry = config_entry
+        self.device_id = device_id
+        self.discovery_data = discovery_data
+        self.hass = hass
+        self._remove_discovery = None
+        self._remove_device_updated = None
+        self._sub_state = None
+        self._value_template = None
+
+        self._setup_from_config(config)
+
+    async def discovery_update(self, payload):
+        """Handle discovery update."""
+        discovery_hash = self.discovery_data[ATTR_DISCOVERY_HASH]
+        _LOGGER.info(
+            "Got update for tag scanner with hash: %s '%s'", discovery_hash, payload
+        )
+        if not payload:
+            # Empty payload: Remove tag scanner
+            _LOGGER.info("Removing tag scanner: %s", discovery_hash)
+            await self.tear_down()
+            if self.device_id:
+                await cleanup_device_registry(self.hass, self.device_id)
+        else:
+            # Non-empty payload: Update tag scanner
+            _LOGGER.info("Updating tag scanner: %s", discovery_hash)
+            config = PLATFORM_SCHEMA(payload)
+            self._config = config
+            if self.device_id:
+                await _update_device(self.hass, self._config_entry, config)
+            self._setup_from_config(config)
+            await self.subscribe_topics()
+
+    def _setup_from_config(self, config):
+        self._value_template = lambda value, error_value: value
+        if CONF_VALUE_TEMPLATE in config:
+            value_template = config.get(CONF_VALUE_TEMPLATE)
+            value_template.hass = self.hass
+
+            self._value_template = value_template.async_render_with_possible_json_value
+
+    async def setup(self):
+        """Set up the MQTT tag scanner."""
+        discovery_hash = self.discovery_data[ATTR_DISCOVERY_HASH]
+        await self.subscribe_topics()
+        if self.device_id:
+            self._remove_device_updated = self.hass.bus.async_listen(
+                EVENT_DEVICE_REGISTRY_UPDATED, self.device_removed
+            )
+        self._remove_discovery = async_dispatcher_connect(
+            self.hass,
+            MQTT_DISCOVERY_UPDATED.format(discovery_hash),
+            self.discovery_update,
+        )
+
+    async def subscribe_topics(self):
+        """Subscribe to MQTT topics."""
+
+        async def tag_scanned(msg):
+            tag_id = self._value_template(msg.payload, error_value="")
```

Comment:
Should we strip here? Now we pass unstripped value to scan tag

Reply:
Good call, fixed.